components/style/properties/longhand/text.mako.rs

```diff
@@ -25,127 +25,15 @@ ${helpers.single_keyword("unicode-bidi",
-                         -> Result<SpecifiedValue, ParseError<'i>> {
-        let mut result = SpecifiedValue::empty();
-        if input.try(|input| input.expect_ident_matching("none")).is_ok() {
-            return Ok(result)
-        }
-        let mut empty = true;
-
-        loop {
-            let result: Result<_, ParseError> = input.try(|input| {
-                let location = input.current_source_location();
-                match input.expect_ident() {
-                    Ok(ident) => {
-                        (match_ignore_ascii_case! { &ident,
-                            "underline" => if result.contains(SpecifiedValue::UNDERLINE) { Err(()) }
-                                           else { empty = false; result.insert(SpecifiedValue::UNDERLINE); Ok(()) },
-                            "overline" => if result.contains(SpecifiedValue::OVERLINE) { Err(()) }
-                                          else { empty = false; result.insert(SpecifiedValue::OVERLINE); Ok(()) },
-                            "line-through" => if result.contains(SpecifiedValue::LINE_THROUGH) { Err(()) }
-                                              else {
-                                                  empty = false;
-                                                  result.insert(SpecifiedValue::LINE_THROUGH); Ok(())
-                                              },
-                            "blink" => if result.contains(SpecifiedValue::BLINK) { Err(()) }
-                                       else { empty = false; result.insert(SpecifiedValue::BLINK); Ok(()) },
-                            _ => Err(())
-                        }).map_err(|()| {
-                            location.new_custom_error(SelectorParseErrorKind::UnexpectedIdent(ident.clone()))
-                        })
-                    }
-                    Err(e) => return Err(e.into())
-                }
-            });
-            if result.is_err() {
-                break;
-            }
-        }
-
-        if !empty { Ok(result) } else { Err(input.new_custom_error(StyleParseErrorKind::UnspecifiedError)) }
-    }
-
-    % if product == "servo":
-        fn cascade_property_custom(_declaration: &PropertyDeclaration,
-                                   context: &mut computed::Context) {
-            longhands::_servo_text_decorations_in_effect::derive_from_text_decoration(context);
-        }
-    % endif
-
-    #[cfg(feature = "gecko")]
-    impl_bitflags_conversions!(SpecifiedValue);
-</%helpers:longhand>
+${helpers.predefined_type("text-decoration-line",
+                    "TextDecorationLine",
+                    "specified::TextDecorationLine::get_initial_value()",
+                    initial_specified_value="specified::TextDecorationLine::get_initial_value()",
+                    custom_cascade= product == 'servo',
+                    cascade_property_custom="specified::TextDecorationLine::cascade_property_custom",
```

Comment:
what do you think about `custom_cascade_function` instead? The `cascade_property_custom` name looks kinda weird.